### PR TITLE
[TG-4345] Parse thrown exceptions in java

### PR DESCRIPTION
**Pull request by jeannielynnmoulton**
* Adds a simple method `get_super_class()` from a java class type. The bytecode was already parsed for this, but never sensibly stored as far as I could tell.
* Parses the exceptions attributes which contains the information after `throws` in `methodName() throws Exception1, Exception2`
* Adds a new type `java_method_typet` which extends `code_typet` so that the methods involving exceptions which are java specific do not pollute `code_typet` (and the way we use it, `java_method_typet` is semantically more correct). I would appreciate suggestions for what else could be moved into this, but will address those in another PR.

This information will be used in test-gen to print the exceptions a test method throws more precisely, i.e., instead of `throws Throwable` it will be `throws SpecificException, AnotherException`, but this is a work in progress yet.

...

test-gen bump: https://github.com/diffblue/test-gen/pull/2094

### CODE REVIEW

File: jbmc/src/java_bytecode/java_types.h [hunk 1]
```diff
@@ -131,6 +131,16 @@ class java_class_typet:public class_typet
     return set(ID_outer_class, outer_class);
   }
 
+  const irep_idt get_super_class() const
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_types.h]: Should be `const irep_idt &` (same for outer_class above, actually)

File: jbmc/src/java_bytecode/java_types.h [hunk 2]
```diff
@@ -131,6 +131,16 @@ class java_class_typet:public class_typet
...
+  {
+    return get(ID_super_class);
+  }
+
+  void set_super_class(irep_idt super_class)
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_types.h]: `const irep_idt &`

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 3]
```diff
@@ -1197,11 +1197,11 @@ void java_bytecode_parsert::rmethod_attribute(methodt &method)
 
       u2 handler_pc=read_u2();
       u2 catch_type=read_u2();
-      method.exception_table[e].start_pc=start_pc;
-      method.exception_table[e].end_pc=end_pc;
-      method.exception_table[e].handler_pc=handler_pc;
+      method.handled_exception_table[e].start_pc=start_pc;
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: clang might complain here

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 4]
```diff
@@ -1655,6 +1658,32 @@ void java_bytecode_parsert::rinner_classes_attribute(
...
 }
 
+/// Corresponds to the element_value structure
+/// Described in Java 8 specification 4.7.5
+/// https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.7.5
+/// Parses the Exceptions attribute for the current method,
+/// which contains a table of the exceptions the method throws.
+void java_bytecode_parsert::rexceptions_attribute(methodt &method, const u4 &attribute_length)
+{
+  std::string name = method.name.c_str();
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Unused variable

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 5]
```diff
@@ -1655,6 +1658,32 @@ void java_bytecode_parsert::rinner_classes_attribute(
...
+    return pool_entry(index);
+  };
+
+  if (number_of_exceptions == 0)
+    return;
+  for(size_t i = 0; i < number_of_exceptions; i++)
+  {
+    u2 exception_index_table=read_u2();
+    pool_entryt pool_entry_temp=pool_entry_lambda(exception_index_table);
+    const irep_idt temp = constant(exception_index_table).type().get(ID_C_base_name);
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: `const irep_idt`

File: src/util/std_types.h [hunk 6]
```diff
@@ -767,6 +767,8 @@ class code_typet:public typet
   class parametert;
   typedef std::vector<parametert> parameterst;
 
+  typedef std::vector<irept> exceptionst;
```
Comment (peterschrammel) [inline src/util/std_types.h]: Shouldn't that be in `java_method_typet`?

Comment (jeannielynnmoulton) [reply src/util/std_types.h]: It is. It's just done in a later commit. If you view all of the files changed, there are no changes to std_types in this PR.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 7]
```diff
@@ -1049,7 +1049,7 @@ codet java_bytecode_convert_methodt::convert_instructions(
        i_it->statement=="monitorexit")))
     {
       const std::vector<unsigned int> handler =
-        try_catch_handler(i_it->address, method.exception_table);
+        try_catch_handler(i_it->address, method.handled_exception_table);
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: I'm not sure about this one. How many exception tables are there?

File: src/util/std_types.h (hunk 6 above)
Comment (LAJW) [reply src/util/std_types.h]: Not a fan of typedefs like this one. It's not much shorter than a fully qualified vector and it gives you less information when it's explicitly specified,

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_class.cpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      load_java_class("ChildClass", "./java_bytecode/java_bytecode_parser");
+    WHEN("Parsing the class file structure")
+    {
+      THEN("We should be able to get the super classes of the hierarchy")
+      {
+        const symbolt &class_symbol1 =
+          new_symbol_table.lookup_ref("java::ChildClass");
+        const java_class_typet java_class1 =
+          to_java_class_type(class_symbol1.type);
+        REQUIRE(java_class1.get_super_class() == "ParentClass");
```
Comment (LAJW) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_class.cpp]: FYI: You can have multiple `THEN` and `WHEN` sections. That way you'll end up with more smaller tests. The benefit (apart from better variable names and shorter test body) is if one of the asserts below fails, all others will still be checked independently.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 6dde5dc89a6cec1ddfac33ccb3b483b174ed403a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79889334
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

File: src/util/std_types.h (hunk 6 above)
Comment (peterschrammel) [reply src/util/std_types.h]: Then these changes should be squashed. Adding code and fixing it up within the same PR is rather confusing.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 9]
```diff
@@ -1655,6 +1658,32 @@ void java_bytecode_parsert::rinner_classes_attribute(
...
+{
+  std::string name = method.name.c_str();
+  u2 number_of_exceptions = read_u2();
+  method.throws_exception_table.resize(number_of_exceptions);
+
+  const auto pool_entry_lambda = [this](u2 index) -> pool_entryt & {
+    return pool_entry(index);
+  };
+
+  if (number_of_exceptions == 0)
```
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Why is this necessary? Surely, if the number of exceptions is 0, then the loop below will loop 0 times and the control flow will leave this function anyway.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 10]
```diff
@@ -1655,6 +1658,32 @@ void java_bytecode_parsert::rinner_classes_attribute(
...
+    return;
+  for(size_t i = 0; i < number_of_exceptions; i++)
+  {
+    u2 exception_index_table=read_u2();
```
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Hint: Needs clang formatting, due to the lack of spaces around the `=`.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 5 above)
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: `temp` isn't a fantastic variable name. I guess this should be something like `exception_name`.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 11]
```diff
@@ -587,6 +587,13 @@ void java_bytecode_convert_methodt::convert(
   method_symbol.location=m.source_location;
   method_symbol.location.set_function(method_identifier);
 
+  code_typet::exceptionst &exceptions_list = code_type.throws_exceptions();
+  for(const auto &p : m.throws_exception_table)
```
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: I dislike using just `p` a variable name. Having the type (if it is short) can also be handy. Maybe `const irep_idt &exception_name` here would be good?

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp [hunk 12]
```diff
@@ -596,4 +597,24 @@ SCENARIO(
...
+      THEN("We should be able to get the list of exceptions it throws")
+      {
+        const symbolt &class_symbol =
+          new_symbol_table.lookup_ref("java::ThrowsExceptions");
+        const symbolt &method_symbol =
+          new_symbol_table.lookup_ref("java::ThrowsExceptions.test:()V");
+        const code_typet method = to_code_type(method_symbol.type);
+        const std::vector<irept> exceptions = method.throws_exceptions();
+        REQUIRE(id2string(exceptions.at(0).id()) == "CustomException");
+        REQUIRE(id2string(exceptions.at(1).id()) == "java.io.IOException");
```
Comment (thomasspriggs) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp]: Can we actually rely on the ordering here and is the ordering actually important? Maybe checking if `exceptions` contains the exception which we are checking for, would be more robust.

File: jbmc/src/java_bytecode/java_types.h [hunk 13]
```diff
@@ -232,6 +242,33 @@ inline bool can_cast_type<java_class_typet>(const typet &type)
...
+
+inline const java_method_typet &to_java_method_type(const typet &type)
+{
+  assert(type.id()==ID_code);
+  return static_cast<const java_method_typet &>(type);
+}
+
+inline java_method_typet &to_java_method_type(typet &type)
+{
+  assert(type.id()==ID_code);
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_types.h]: Should java_method_typet have its own `ID_` tag?

Use invariants/precondition over asserts. Asserts are usually disabled in release builds.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 14]
```diff
@@ -1655,6 +1658,32 @@ void java_bytecode_parsert::rinner_classes_attribute(
...
+    pool_entryt pool_entry_temp=pool_entry_lambda(exception_index_table);
+    const irep_idt temp = constant(exception_index_table).type().get(ID_C_base_name);
+    method.throws_exception_table.push_back(temp);
+  }
+}
+
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: There are so many unused variables, the body of this function could be rewritten as:

```c++
const u2 number_of_exceptions = read_u2(); 
for(size_t i = 0; i < number_of_exceptions; i++)
  method.throws_exception_table.push_back(
    constant(read_u2()).type().get(ID_C_base_name));
```

I'm not sure about `.resize()` - do you want to shrink the vector, fill it with empty elements or [`.reserve()`](http://www.cplusplus.com/reference/vector/vector/reserve/) elements to make push_back more efficient (in which case it's not worth it here I think)?

Rather than passing the method as an argument, you could return a vetor of constants. Right now it's not clear whether you're appending exceptions to an existing vector or creating new one from scratch. This forces the reader to look elsewhere whether this vector contains anything or not.

Also, this function doesn't seem to modify the state, so could be marked as `const`, but for that you'd have to mark `read_u2` and `read_bytes` as const (I tried that, and it compiles).

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 15]
```diff
@@ -587,6 +587,13 @@ void java_bytecode_convert_methodt::convert(
...
+  {
+    if(!p.empty())
+      exceptions_list.push_back(irept(p));
```
Comment (LAJW) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: The disadvantage of `typedef` comes into play here. If they weren't used, it would've never occurred to you to convert `irept` to `irept`.

This function could be rewritten as `std::copy_if`.

**Review by LAJW — COMMENTED** [review]
This `resize` bit is dubious. I'd like to see tests regrouped to allow more specific failures.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 16]
```diff
@@ -1655,6 +1658,32 @@ void java_bytecode_parsert::rinner_classes_attribute(
...
+/// Corresponds to the element_value structure
+/// Described in Java 8 specification 4.7.5
+/// https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.7.5
+/// Parses the Exceptions attribute for the current method,
+/// which contains a table of the exceptions the method throws.
+void java_bytecode_parsert::rexceptions_attribute(methodt &method, const u4 &attribute_length)
+{
+  std::string name = method.name.c_str();
+  u2 number_of_exceptions = read_u2();
+  method.throws_exception_table.resize(number_of_exceptions);
```
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: I think this is actually introducing problems which you have to deal with later.

If my understanding of how this `rexceptions_attribute` function is working is correct, then it is creating `number_of_exceptions` empty entries in the `throws_exception_table` and then adding the actual exceptions after this. Resulting in a vector which is twice as long as you expect, which contains junk empty items.

If you delete this line, then there will be no empty exceptions in the vector and you won't need to check for empty names when you get them out later.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 17]
```diff
@@ -1655,6 +1658,32 @@ void java_bytecode_parsert::rinner_classes_attribute(
...
+
+  const auto pool_entry_lambda = [this](u2 index) -> pool_entryt & {
```
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: I think this `pool_entry_lambda` is making the code harder to read. Why not use `pool_entry(exception_index_table)` inline, instead of `pool_entry_lambda(exception_index_table)`?

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 7 above)
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: There are two. There this one which deals with exceptions that can be caught. And the one that I am parsing in this PR which are exceptions that a method throws. Strictly speaking, in the bytecode, this is called exception_table and mine is also called exception_table. I thought this just made it a bit clearer what is what.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 17 above)
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: I've done this another way, so this is no longer required.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 5 above)
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: oversight.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 14 above)
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: I forgot to delete the pool entry when I used constant, so this is resolved.

I've taken your advice of returning a vector. Everything else is a void method, so it ruins the beauty of the declarations.

ALL the things are const.

File: jbmc/src/java_bytecode/java_types.h (hunk 13 above)
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_types.h]: Not sure. It's not required here. ID_code is what is intended.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 14 above)
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Actually, not everything is const. There were some methods further down the chain that were required to be const and couldn't be.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 00c51890ae865c1beae4457820dd7fe72181dbed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80118123
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by thomasspriggs — APPROVED** [review]
LGTM

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 07b5faaa502ec491099433a2a9eeaf8f1f2e5036).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80122234
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 7cf4d1e22ade75fb77a008dc52ebe2b35da80473).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80126073

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: b4e3609d21502ab0c15e0833905fe56ccc5f1597).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80199423

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 93f970841508a22f5e103978265bb06ff3294a51).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80285131

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 18]
```diff
@@ -151,7 +152,7 @@ class java_bytecode_parsert:public parsert
     }
   }
 
-  u8 read_bytes(size_t bytes)
+  u8 read_bytes(size_t bytes) const
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Not entirely sure about adding `const` here (is it in right commit? This class is responsible for reading a file, read_bytes advances the "head" of where we're reading from. Don't understand why `in->get()` is const

Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Ah I see this was under the advice of @LAJW - feel free to disregard my comment, me and Lukasz can discuss and reach a conclusion.

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp [hunk 19]
```diff
@@ -596,4 +597,32 @@ SCENARIO(
...
+    {
+      THEN("We should be able to get the list of exceptions it throws")
+      {
+        const symbolt &class_symbol =
+          new_symbol_table.lookup_ref("java::ThrowsExceptions");
+        const symbolt &method_symbol =
+          new_symbol_table.lookup_ref("java::ThrowsExceptions.test:()V");
+        const code_typet method = to_code_type(method_symbol.type);
+        const std::vector<irept> exceptions = method.throws_exceptions();
+        REQUIRE(exceptions.size() == 2);
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp]: Nit: you can use `Catch:Matchers::Vector` to do this in a nicer way - when I have an IDE I can show you a snippet, else you can search the code base for use cases. Not blocking but makes unit test and errors from it nicer to read.

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp [hunk 20]
```diff
@@ -596,4 +597,32 @@ SCENARIO(
...
+            exceptions.begin(), exceptions.end(), irept("CustomException")) !=
+          exceptions.end());
+        REQUIRE(
+          std::find(
+            exceptions.begin(),
+            exceptions.end(),
+            irept("java.io.IOException")) != exceptions.end());
+      }
+    }
+  }
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp]: Please add a test for empty exception list

File: jbmc/src/java_bytecode/java_types.h [hunk 21]
```diff
@@ -242,6 +242,32 @@ inline bool can_cast_type<java_class_typet>(const typet &type)
   return can_cast_type<class_typet>(type);
 }
 
+class java_method_typet : public code_typet
+{
+public:
+  const std::vector<irept> &throws_exceptions() const
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_types.h]: I strongly dislike returning `vector<irept>` as caller knows literaly nothing about what those `irept`s might be, could be `symbol_typet` as in `bases()` on a class, or just strings as they actually are. To hide the implementation detail, I suggest changing to:

```C++
const std::vector<irep_idt> &throws_exceptions() const
{
  std:vector<irep_idt>exceptions;
  std::transform(
    find(ID_exceptions_thrown_list).get_sub().being(), 
    find(ID_exceptions_thrown_list).get_sub().end(), 
    irept::idt);
  return exceptions;
}
```

There may (must??) be a nicer way of doing this without the copy, but to me even this is preferable to an interface that returns essentially a typeless array. 

I'd also try and get rid of the setter, and instead take a `std::vector<irep_idt>` in the constructor so it is always in a valid state.

File: jbmc/src/java_bytecode/java_types.h [hunk 22]
```diff
@@ -242,6 +242,32 @@ inline bool can_cast_type<java_class_typet>(const typet &type)
...
+
+  std::vector<irept> &throws_exceptions()
+  {
+    return add(ID_exceptions_thrown_list).get_sub();
+  }
+};
+
+inline const java_method_typet &to_java_method_type(const typet &type)
+{
+  PRECONDITION(type.id() == ID_code);
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_types.h:264]: To add a bit more type safety, you might want to add a `ID_C_java_method_type` so a caller can be sure that it really is a `java_method_typet` not a `code_typet`.

File: jbmc/src/java_bytecode/java_types.h (hunk 21 above)
Comment (thk123) [reply jbmc/src/java_bytecode/java_types.h]: OK I see the constructor part is not possible without some restructuring. Instead I suggest a `add_throws_exception` method that adds to the sub, since that is the only way anyone should be modifying it. (by giving the whole sub tree away, any caller could inadvertently remove all the exceptions).

**Review by thk123 — COMMENTED** [review]
I'd really like to see the interface of `java_method_typet` changed before this gets merged if it isn't too hard.

File: jbmc/src/java_bytecode/java_types.h (hunk 22 above)
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/java_types.h:264]: Why would this be a comment with `C_` in the name instead of `ID_java_method_type`?

Comment (thk123) [reply jbmc/src/java_bytecode/java_types.h:264]: Consistency with other types, which isn't a great reason. I think you're probably right in fact, if I have a C style function pointer that probably shouldn't be treated equivalently to a static java function that doesn't throw any exceptions.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 18 above)
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: I've removed them for now because the function that I was trying to make `const` could not be `const` anyway. This was a byproduct of that and I left it in because it seemed sensible, but I don't want to muddy it up if there is uncertainty. It's not required for my PR.

File: jbmc/src/java_bytecode/java_types.h [hunk 23]
```diff
@@ -232,6 +242,35 @@ inline bool can_cast_type<java_class_typet>(const typet &type)
...
+public:
+  const std::vector<irep_idt> throws_exceptions() const
+  {
+    std::vector<irep_idt> exceptions;
+    for(const auto &e : find(ID_exceptions_thrown_list).get_sub())
+      exceptions.push_back(e.id());
+    return exceptions;
+  }
+
+  void add_throws_exceptions(irept exception)
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_types.h]: Make this an `irep_idt` to completely hide the `irept` from the public interface.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 66ba3db5b9bde9aa57539a7f508c3922a518b355).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80483614
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp [hunk 24]
```diff
@@ -596,4 +597,45 @@ SCENARIO(
...
+          new_symbol_table.lookup_ref("java::ThrowsExceptions.test:()V");
+        const java_method_typet method =
+          to_java_method_type(method_symbol.type);
+        const std::vector<irep_idt> exceptions = method.throws_exceptions();
+        REQUIRE(exceptions.size() == 2);
+        REQUIRE(
+          std::find(
+            exceptions.begin(),
+            exceptions.end(),
+            irept("CustomException").id()) != exceptions.end());
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp:619]: No need to construct an irept and then get the `id()`

**Review by thk123 — APPROVED** [review]
A thing to bear in mind in future, every symbol has two names, (e.g. `java.io.Exception` vs. `java::java.io.Exception`) 

When you return an `irep_idt` it isn't obvious which one you're getting back. I think leave for now, but it might be better to return `symbol_typet` which implies you can look it up in the symbol table (and therefore definitely the `java::java.io.Exception`. We've not been consistent about this in the past, but I think we should try to be in future.

**Comment by thk123** [comment]
CI failure looked like a glitch so restarted.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 655248a0bd3aeb05fa24e569773da2362f8756a8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80493176

File: jbmc/src/java_bytecode/java_types.h (hunk 22 above)
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_types.h:264]: This has been added in https://github.com/diffblue/cbmc/pull/2661
